pkg/operator/k8sutil/pod.go

```diff
@@ -296,13 +297,17 @@ func deleteResourceAndWait(namespace, name, resourceType string,
 	return fmt.Errorf("gave up waiting for %s pods to be terminated", name)
 }
 
-// Environment variables used by storage cluster daemons
+// ClusterDaemonEnvVars Environment variables used by storage cluster daemon
 func ClusterDaemonEnvVars(image string) []v1.EnvVar {
 	return []v1.EnvVar{
 		{Name: "CONTAINER_IMAGE", Value: image},
 		{Name: "POD_NAME", ValueFrom: &v1.EnvVarSource{FieldRef: &v1.ObjectFieldSelector{FieldPath: "metadata.name"}}},
 		{Name: "POD_NAMESPACE", ValueFrom: &v1.EnvVarSource{FieldRef: &v1.ObjectFieldSelector{FieldPath: "metadata.namespace"}}},
 		{Name: "NODE_NAME", ValueFrom: &v1.EnvVarSource{FieldRef: &v1.ObjectFieldSelector{FieldPath: "spec.nodeName"}}},
+		{Name: "POD_MEMORY_LIMIT", ValueFrom: &v1.EnvVarSource{ResourceFieldRef: &v1.ResourceFieldSelector{Resource: "limits.memory"}}},     // Bytes
```

Comment:
Some of them will, for instance `memory.limit` will be the host total memory but `memory.request` will remain 0.
Yes, it's confusing because the value is actually 0 when we check in the operator code but at instantiation time k8s populates it with a value, host total memory (read from `/proc/meminfo`).

Reply:
why set a limit like "host total memory"?  The pod will never ever be able to reach this limit.  On the other hand, if you use a memory CGroup limit that is too low, then you'll have a memory allocation failure and the pod will crash.  The purpose for a memory CGroup limit in Kubernetes is to ensure that a pod does not use more memory than Kubernetes thinks it needs.   Just my opinion: make the memory CGroup limit  higher than the limit passed to the Ceph daemon in ceph.conf, if we hit the CGroup limit under this circumstance then it is clearly a Ceph issue, since Ceph failed to live up to its obligation to restrict its memory consumption for a daemon in this pod, make sense?  If so, here's a possible formula:  CGroup memory limit = 1 GB + 1.1 x daemon-limit, where daemon limit is the ceph.conf limit for that particular daemon in the pod.  Example: for MDS pod, mds_cache_memory_limit is the ceph.conf limit, suppose this was 60 GB, so memory CGroup limit would be 1 GB + this value x 110% = 67 GB, and we'd want to inform Kubernetes that this pod requires this much memory.     We'd potentially be wasting 7 GB  or ~10% memory but perhaps this is worth it to avoid a daemon crash and allow Kubernetes scheduler to do its job?

Reply:
We don't set it, Kubernetes does, we don't have any control on that behavior. I share your concern @bengland2 about POD being terminated though. I guess what we could do is intrust Ceph rely on `memory.request` so that leaves some memory (limit - request). However, if the POD dies because of OOM, it'll be difficult to tell if this was because of a backfilling operation happening or because BS cache leaked or both...

Reply:
I think I share the concerns @bengland2 has. In my head, I view `memory.request` as "this is the amount of memory that is used under normal conditions for my application" and view `memory.limit` as "if my application uses this much memory, there is a resource leak, and k8s is free to restart the pod so that other processes on the node don't run out of mem because of this one pod.

Nit: Kubernetes refers to pods as lower case. I have been thrown off a few times seeing `POD` and thinking it is an acronym for something else.

Reply:
This discussion should continue on ceph-devel in the thread with subject "Ceph, container and memory".

Reply:
I revisited this conversation and think I may have missed some nuances the first time around. I think I understand better @leseb's comments. Seb, it might be good to mention in a code comment somewhere that kubernetes' behavior is to populate the limits with system values if Rook or the user doesn't set any.